### PR TITLE
DOC: optimize: add comparison of optimizers to guide

### PR DESCRIPTION
**Pull request by koko1928**
#### Reference issue

Toward #17404
Supersedes gh-17491

#### What does this implement/fix?

Added optimization solvers reference tables and its description.

### CODE REVIEW

File: doc/source/tutorial/optimize.rst [hunk 1]
```diff
@@ -1106,6 +1106,183 @@ For example, to find the minimum of :math:`J_{1}\left( x \right)` near
...
+     - ✓
+     - 
+     - 
+   * - basinhopping
+     - (✓)
+     - (✓)
+     - 
+     - 
+     - 
+(✓): Depending on the chosen local minimizer
```
Comment (tylerjereddy) [inline doc/source/tutorial/optimize.rst:1272]: I guess you have a minor doc build issue in CI: `/home/circleci/repo/doc/source/tutorial/optimize.rst:1272: WARNING: Explicit markup ends without a blank line; unexpected unindent.`

There are a few places in the first paragraph above where spaces are missing around punctuation, but that's pretty minor too.

Comment (koko1928) [reply doc/source/tutorial/optimize.rst]: Thank you for your review. Fixed.

**Comment by tylerjereddy** [comment]
Looks like doc build is still failing--this small diff fixes it for me locally:

<details>

```diff
--- a/doc/source/tutorial/optimize.rst
+++ b/doc/source/tutorial/optimize.rst
@@ -1269,6 +1269,7 @@ which is especially important when the objective function has multiple local min
      - 
      - 
      - 
+
 (✓) = Depending on the chosen local minimizer
```

</details>

For the expert reviewers, a sample render of the tables in this PR is below the fold, in the absence of the passing doc build in CI.

<details>

image
image

</details>

And for reference, the relevant table in the SciPy `1.0` paper is here: https://www.nature.com/articles/s41592-019-0686-2/tables/1

I'll leave most of the inspection to the experts, but one question I have--why does Newton-CG have a "global" checkmark in the published paper, but is listed under local here? The paper does note:

> this is not a guarantee of convergence to a global minimum.

Is that the distinction here, or? Anyway, if the difference confuses me a little, maybe worth noting, but I wouldn't change much until you get feedback from the `optimize` gurus.

**Comment by koko1928** [comment]
> Looks like doc build is still failing--this small diff fixes it for me locally:
> 
> For the expert reviewers, a sample render of the tables in this PR is below the fold, in the absence of the passing doc build in CI.
> 
> And for reference, the relevant table in the SciPy `1.0` paper is here: https://www.nature.com/articles/s41592-019-0686-2/tables/1
> 
> I'll leave most of the inspection to the experts, but one question I have--why does Newton-CG have a "global" checkmark in the published paper, but is listed under local here? The paper does note:
> 
> > this is not a guarantee of convergence to a global minimum.
> 
> Is that the distinction here, or? Anyway, if the difference confuses me a little, maybe worth noting, but I wouldn't change much until you get feedback from the `optimize` gurus.

I am not an expert in this field so I don't know the details, but it is based on the spreadsheet presented by @dschmitz89  in DOC: comparison of optimizers in optimization guide #17404.

**Comment by rkern** [comment]
As described in the footnote, "global convergence" is a very specific property mostly unrelated to being a "global optimizer". None of the methods in that table in the paper are global optimizers.

**Comment by dschmitz89** [comment]
The tables look great, thank you!

My only concern is where they should be placed. I would insert them in the beginning of the local or global optimization tutorials each. Another option would be to place them together at the top of the whole tutorial page. Opinions?

**Comment by koko1928** [comment]
> The tables look great, thank you!
> 
> My only concern is where they should be placed. I would insert them in the beginning of the local or global optimization tutorials each. Another option would be to place them together at the top of the whole tutorial page. Opinions?

I think it is a good idea to list it at the top.

**Comment by koko1928** [comment]
> > The tables look great, thank you!
> > My only concern is where they should be placed. I would insert them in the beginning of the local or global optimization tutorials each. Another option would be to place them together at the top of the whole tutorial page. Opinions?
> 
> I think it is a good idea to list it at the top.

What changes should I make?

**Comment by dschmitz89** [comment]
Since we are 

> > > The tables look great, thank you!
> > > My only concern is where they should be placed. I would insert them in the beginning of the local or global optimization tutorials each. Another option would be to place them together at the top of the whole tutorial page. Opinions?
> > 
> > 
> > I think it is a good idea to list it at the top.
> 
> What changes should I make?

Since we are both happy with moving it to the top, let's do so :).

File: doc/source/tutorial/optimize.rst [hunk 2]
```diff
@@ -16,6 +16,184 @@ optimization algorithms. A detailed listing is available:
...
+     - 
+   * - Newton-CG
+     - ✓
+     - ✓
+     - 
+     - 
+     - ✓
+   * - dogleg
+     - ✓
+     - must be pos. def.
```
Comment (dschmitz89) [inline doc/source/tutorial/optimize.rst:86]:
```suggestion
     - (✓)
```

File: doc/source/tutorial/optimize.rst [hunk 3]
```diff
@@ -16,6 +16,184 @@ optimization algorithms. A detailed listing is available:
...
+     - ✓
+     - ✓
+   * - SLSQP
+     - ✓
+     - 
+     - ✓
+     - ✓
+     - 
+
+
```
Comment (dschmitz89) [inline doc/source/tutorial/optimize.rst:133]:
```suggestion

(✓) = must be positive definite

```

File: doc/source/tutorial/optimize.rst [hunk 4]
```diff
@@ -16,6 +16,184 @@ optimization algorithms. A detailed listing is available:
...
 
 
+Optimization Solvers Reference
+------------------------------
+
+Provide an overview of various optimization solvers,which are tools 
+used to find the minimum or maximum of an objective function subject 
+to constraints. These solvers are divided into two main categories: 
+Local Optimization and Global Optimization.Each category has a set of 
+solvers with unique characteristics and capabilities.
```
Comment (dschmitz89) [inline doc/source/tutorial/optimize.rst:26]:
```suggestion
This section provides an overview of the optimization solvers
available in SciPy to assist users  infinding a suitable solver for a given
problem. These solvers are divided into two main categories: 
local optimization and global optimization. Each category has a set of 
solvers with unique characteristics and capabilities.
```

**Review by dschmitz89 — DISMISSED** [review]

**Comment by dschmitz89** [comment]
Sorry for the back and forth but I am having second thoughts looking at the optimization user guide again. For users of linear programming or MILP for example, it would be confusing to be greeted in the second paragraph with tables of nonlinear optimizers. What about moving the tables to the dedicated sections for local and global optimization?

**Comment by mdhaber** [comment]
> For users of linear programming or MILP for example, it would be confusing to be greeted in the second paragraph with tables of nonlinear optimizers.

Maybe there would be some advice about choosing between LP, NLP, global, curve fitting, root finding, etc... and linking to the relevant sections of the guide? But that opens a larger can of worms, so I'd understand if it's not something you want to tackle here.

**Comment by koko1928** [comment]
> Sorry for the back and forth but I am having second thoughts looking at the optimization user guide again. For users of linear programming or MILP for example, it would be confusing to be greeted in the second paragraph with tables of nonlinear optimizers. What about moving the tables to the dedicated sections for local and global optimization?

Yes. That might be easier to understand.

**Comment by koko1928** [comment]
> > For users of linear programming or MILP for example, it would be confusing to be greeted in the second paragraph with tables of nonlinear optimizers.
> 
> Maybe there would be some advice about choosing between LP, NLP, global, curve fitting, root finding, etc... and linking to the relevant sections of the guide? But that opens a larger can of worms, so I'd understand if it's not something you want to tackle here.
> 
> Table looks good, though. Great to finally have it, since it's been an idea for a long time. What do you think about having a different mark for _requiring_ the user to specify the derivatives (like some of the trust methods)?

Good idea. If necessary, you can request a code change and I can reflect it.

**Comment by dschmitz89** [comment]
> > For users of linear programming or MILP for example, it would be confusing to be greeted in the second paragraph with tables of nonlinear optimizers.
> 
> Maybe there would be some advice about choosing between LP, NLP, global, curve fitting, root finding, etc... and linking to the relevant sections of the guide? But that opens a larger can of worms, so I'd understand if it's not something you want to tackle here.

Yep, exactly, that would quickly spiral this PR out of control.

> Table looks good, though. Great to finally have it, since it's been an idea for a long time. What do you think about having a different mark for _requiring_ the user to specify the derivatives (like some of the trust methods)?

Good idea. I have never used those trust methods though myself, will have to dig to see which ones require the Hessian. I had always assumed that all our optimizers can fall back to finite differences.

**Comment by koko1928** [comment]
>What do you think about having a different mark for requiring the user to specify the derivatives

As soon as this is implemented, it would be good to Merge this PR and start a new issue.

>Maybe there would be some advice about choosing between LP, NLP, global, curve fitting, root finding, etc... and linking to the relevant sections of the guide

This idea is very good, but I think it is best discussed in a new Issue or PR.

**Comment by koko1928** [comment]
Is there anything else we need to do for this PR to be merged?

**Comment by dschmitz89** [comment]
@koko1928 : sorry for the late reply. Could you move the tables to the beginning of the local and global optimization sections respectively? After that, I will make a pass over the description and merge.

**Comment by koko1928** [comment]
> @koko1928 : sorry for the late reply. Could you move the tables to the beginning of the local and global optimization sections respectively? After that, I will make a pass over the description and merge.

Previously, I made a change to place it at the top for clarity.Will I change the placement again?

**Comment by dschmitz89** [comment]
> > @koko1928 : sorry for the late reply. Could you move the tables to the beginning of the local and global optimization sections respectively? After that, I will make a pass over the description and merge.
> 
> Previously, I made a change to place it at the top for clarity.Will I change the placement again?

Sorry for this back and fourth. It unfortunately happens sometimes in reviews and I understand that it can be annoying.

I am now convinced that placing the tables in the beginnings of the respective sections is the better choice, so I am not gonna ask to move them again.

**Comment by koko1928** [comment]
> > > @koko1928 : sorry for the late reply. Could you move the tables to the beginning of the local and global optimization sections respectively? After that, I will make a pass over the description and merge.
> > 
> > 
> > Previously, I made a change to place it at the top for clarity.Will I change the placement again?
> 
> Sorry for this back and fourth. It unfortunately happens sometimes in reviews and I understand that it can be annoying.
> 
> I am now convinced that placing the tables in the beginnings of the respective sections is the better choice, so I am not gonna ask to move them again once they are there.

I agree with putting a table at the beginning of each section. However, as far as the current tutorial is concerned, there is a section for global optimizations, but nothing for local optimizations. If you want to add a section for local optimization, you should consider where you want to add it.

**Comment by dschmitz89** [comment]
> > > > @koko1928 : sorry for the late reply. Could you move the tables to the beginning of the local and global optimization sections respectively? After that, I will make a pass over the description and merge.
> > > 
> > > 
> > > Previously, I made a change to place it at the top for clarity.Will I change the placement again?
> > 
> > 
> > Sorry for this back and fourth. It unfortunately happens sometimes in reviews and I understand that it can be annoying.
> > I am now convinced that placing the tables in the beginnings of the respective sections is the better choice, so I am not gonna ask to move them again once they are there.
> 
> I agree with putting a table at the beginning of each section. However, as far as the current tutorial is concerned, there is a section for global optimizations, but nothing for local optimizations. If you want to add a section for local optimization, you should consider where you want to add it.

Right, I did not look in the details there. How about we add the first three main points in the user guide under a new heading "Local optimization" and put the table in the beginning?

**Comment by koko1928** [comment]
> > > > > @koko1928 : sorry for the late reply. Could you move the tables to the beginning of the local and global optimization sections respectively? After that, I will make a pass over the description and merge.
> > > > 
> > > > 
> > > > Previously, I made a change to place it at the top for clarity.Will I change the placement again?
> > > 
> > > 
> > > Sorry for this back and fourth. It unfortunately happens sometimes in reviews and I understand that it can be annoying.
> > > I am now convinced that placing the tables in the beginnings of the respective sections is the better choice, so I am not gonna ask to move them again once they are there.
> > 
> > 
> > I agree with putting a table at the beginning of each section. However, as far as the current tutorial is concerned, there is a section for global optimizations, but nothing for local optimizations. If you want to add a section for local optimization, you should consider where you want to add it.
> 
> Right, I did not look in the details there. How about we add the first three main points in the user guide under a new heading "Local optimization" and put the table in the beginning?

agree. I think it is possible to add each table to a section, but there is also the issue of where to put the usage instructions. Another option is to refer to the instructions above for proper use of each section.

**Comment by dschmitz89** [comment]
Usage instruc

> > > > > > @koko1928 : sorry for the late reply. Could you move the tables to the beginning of the local and global optimization sections respectively? After that, I will make a pass over the description and merge.
> > > > > 
> > > > > 
> > > > > Previously, I made a change to place it at the top for clarity.Will I change the placement again?
> > > > 
> > > > 
> > > > Sorry for this back and fourth. It unfortunately happens sometimes in reviews and I understand that it can be annoying.
> > > > I am now convinced that placing the tables in the beginnings of the respective sections is the better choice, so I am not gonna ask to move them again once they are there.
> > > 
> > > 
> > > I agree with putting a table at the beginning of each section. However, as far as the current tutorial is concerned, there is a section for global optimizations, but nothing for local optimizations. If you want to add a section for local optimization, you should consider where you want to add it.
> > 
> > 
> > Right, I did not look in the details there. How about we add the first three main points in the user guide under a new heading "Local optimization" and put the table in the beginning?
> 
> agree. I think it is possible to add each table to a section, but there is also the issue of where to put the usage instructions. Another option is to refer to the instructions above for proper use of each section.

For a first mergeable version, let's just leave out the usage instructions then?

**Comment by koko1928** [comment]
I'm getting an error like the one below. How should I do it?

>ERROR: No matching distribution found for jupyterlite-sphinx>=0.13.0

**Comment by dschmitz89** [comment]
> I'm getting an error like the one below. How should I do it?
> 
> >ERROR: No matching distribution found for jupyterlite-sphinx>=0.13.0

This was fixed in main by now. If you merge main into your branch, CI should become green.

**Comment by koko1928** [comment]
> > I'm getting an error like the one below. How should I do it?
> > > ERROR: No matching distribution found for jupyterlite-sphinx>=0.13.0
> 
> This was fixed in main by now. If you merge main into your branch, CI should become green.

I got it. thank you.
The problem has been fixed and it works correctly.

**Comment by koko1928** [comment]
> > > I'm getting an error like the one below. How should I do it?
> > > > ERROR: No matching distribution found for jupyterlite-sphinx>=0.13.0
> > 
> > 
> > This was fixed in main by now. If you merge main into your branch, CI should become green.
> 
> I got it. thank you. The problem has been fixed and it works correctly.

Is there anything else we need to do for this PR to be merged?

**Comment by dschmitz89** [comment]
@andyfaff : would you have time to chime in? The actual tables for the optimizers look good imo but we are having a hard time figuring out a good way to place them into the tutorial page.

File: doc/source/tutorial/optimize.rst [hunk 5]
```diff
@@ -916,6 +915,169 @@ We'll now plot all found minima on a heatmap of the function::
    :alt: "This X-Y plot is a heatmap with the Z value denoted with the lowest points as black and the highest values as white. The image resembles a chess board rotated 45 degrees but heavily smoothed. A red dot is located at many of the minima on the grid resulting from the SHGO optimizer. SHGO shows the global minima as a red X in the top right. A local minima found with dual annealing is a white circle marker in the top left. A different local minima found with basinhopping is a yellow marker in the top center. The code is plotting the differential evolution result as a cyan circle, but it is not visible on the plot. At a glance it's not clear which of these valleys is the true global minima."
    :include-source: 0
 
+Global Optimization Solvers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Global optimization solvers aim to find the global minimum over the entire search space,
+which is especially important when the objective function has multiple local minima.
+
+**Global Optimization Solvers Table:**
```
Comment (andyfaff) [inline doc/source/tutorial/optimize.rst:924]:
```suggestion
**Comparison of Global Optimizers:**
```

File: doc/source/tutorial/optimize.rst [hunk 6]
```diff
@@ -916,6 +915,169 @@ We'll now plot all found minima on a heatmap of the function::
    :alt: "This X-Y plot is a heatmap with the Z value denoted with the lowest points as black and the highest values as white. The image resembles a chess board rotated 45 degrees but heavily smoothed. A red dot is located at many of the minima on the grid resulting from the SHGO optimizer. SHGO shows the global minima as a red X in the top right. A local minima found with dual annealing is a white circle marker in the top left. A different local minima found with basinhopping is a yellow marker in the top center. The code is plotting the differential evolution result as a cyan circle, but it is not visible on the plot. At a glance it's not clear which of these valleys is the true global minima."
    :include-source: 0
 
+Global Optimization Solvers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Global optimization solvers aim to find the global minimum over the entire search space,
```
Comment (andyfaff) [inline doc/source/tutorial/optimize.rst:921]: remove L921 and L922, this info is duplicated from elsewhere.

File: doc/source/tutorial/optimize.rst [hunk 7]
```diff
@@ -916,6 +915,169 @@ We'll now plot all found minima on a heatmap of the function::
...
+------------------
+
+Local Optimization Solvers
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Local optimization solvers are algorithms that search for a local minimum, 
+starting from an initial point. These solvers are particularly useful 
+when the problem is presumed to have a smooth landscape around the initial guess.
+
+**Local Optimization Solvers Table:**
```
Comment (andyfaff) [inline doc/source/tutorial/optimize.rst:981]:
```suggestion
**Local Optimization Solver comparison:**
```

File: doc/source/tutorial/optimize.rst [hunk 8]
```diff
@@ -916,6 +915,169 @@ We'll now plot all found minima on a heatmap of the function::
...
+     - (✓)
+     - 
+     - 
+     - 
+
+(✓) = Depending on the chosen local minimizer
+
+
+
+Local optimization
```
Comment (andyfaff) [inline doc/source/tutorial/optimize.rst:971]: remove L971, L972

File: doc/source/tutorial/optimize.rst [hunk 9]
```diff
@@ -916,6 +915,169 @@ We'll now plot all found minima on a heatmap of the function::
    :alt: "This X-Y plot is a heatmap with the Z value denoted with the lowest points as black and the highest values as white. The image resembles a chess board rotated 45 degrees but heavily smoothed. A red dot is located at many of the minima on the grid resulting from the SHGO optimizer. SHGO shows the global minima as a red X in the top right. A local minima found with dual annealing is a white circle marker in the top left. A different local minima found with basinhopping is a yellow marker in the top center. The code is plotting the differential evolution result as a cyan circle, but it is not visible on the plot. At a glance it's not clear which of these valleys is the true global minima."
    :include-source: 0
 
+Global Optimization Solvers
```
Comment (andyfaff) [inline doc/source/tutorial/optimize.rst:918]:
```suggestion
Comparison of Optimizer features
----------------------------------

Global Optimizers
```

File: doc/source/tutorial/optimize.rst [hunk 10]
```diff
@@ -916,6 +915,169 @@ We'll now plot all found minima on a heatmap of the function::
...
+     - 
+
+(✓) = Depending on the chosen local minimizer
+
+
+
+Local optimization
+------------------
+
+Local Optimization Solvers
```
Comment (andyfaff) [inline doc/source/tutorial/optimize.rst:974]:
```suggestion
Local Optimizers
```

File: doc/source/tutorial/optimize.rst [hunk 11]
```diff
@@ -916,6 +915,169 @@ We'll now plot all found minima on a heatmap of the function::
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Local optimization solvers are algorithms that search for a local minimum,
```
Comment (andyfaff) [inline doc/source/tutorial/optimize.rst:977]:
```suggestion
Local optimizers are algorithms that search for a local minimum, 
```

Comment (andyfaff) [reply doc/source/tutorial/optimize.rst:977]: I think this description can be improved.
They're efficient at finding minima, but can get stuck in local minima. The minima that they find can depend on the point the optimizer starts from.

File: doc/source/tutorial/optimize.rst [hunk 12]
```diff
@@ -916,6 +915,169 @@ We'll now plot all found minima on a heatmap of the function::
...
+
+**Local Optimization Solvers Table:**
+
+.. list-table::
+   :widths: 15 15 15 20 20 15
+   :header-rows: 1
+
+   * - Solver
+     - Uses Gradient
+     - Uses Hessian
```
Comment (andyfaff) [inline doc/source/tutorial/optimize.rst:989]: This table could be fleshed out more w.r.t gradient/hessian information. e.g. LBFGSB uses a gradient, but that gradient can be estimated by the numerical differentiation, OR the user can supply a gradient function.
In comparison Newton-CG definitely requires a gradient function, but can estimate hessian by numerical differentiation or hessian function.
Yet again trust-constr can estimate both gradient and hessian by functions, or by numdiff for gradient and HessianUpdateStrategy for the Hessian.

i.e. there is lots of complexity in what each function can use.

Comment (dschmitz89) [reply doc/source/tutorial/optimize.rst:992]: What about a second table for details about derivatives and hessians?

Comment (koko1928) [reply doc/source/tutorial/optimize.rst:992]: > What about a second table for details about derivatives and hessians?

@andyfaff

**Comment by tylerjereddy** [comment]
It looks like there are a few reviewer comments still left to be addressed. I may bump the milestone for that reason, but could we merge this soon? It looks like the doc PR has been open for almost 4 months now with somewhat-reasonable agreement on the table being "ok" modulo a few nitpicks/follow-ups. Let's squash merge when ready, don't need 34+ commits.

**Comment by koko1928** [comment]
> It looks like there are a few reviewer comments still left to be addressed. I may bump the milestone for that reason, but could we merge this soon? It looks like the doc PR has been open for almost 4 months now with somewhat-reasonable agreement on the table being "ok" modulo a few nitpicks/follow-ups. Let's squash merge when ready, don't need 34+ commits.

If there is no problem, I would like you to merge it.

**Comment by mdhaber** [comment]
I made some changes that I think makes this mergeable. Should have pushed with `[docs only]` but oh well. If the render looks good (and if the table looks correct - i need to check the local minimizer table), I'll merge.  COBYQA can be added in a separate PR.

File: doc/source/tutorial/optimize.rst [hunk 13]
```diff
@@ -916,6 +1006,50 @@ We'll now plot all found minima on a heatmap of the function::
    :alt: "This X-Y plot is a heatmap with the Z value denoted with the lowest points as black and the highest values as white. The image resembles a chess board rotated 45 degrees but heavily smoothed. A red dot is located at many of the minima on the grid resulting from the SHGO optimizer. SHGO shows the global minima as a red X in the top right. A local minima found with dual annealing is a white circle marker in the top left. A different local minima found with basinhopping is a yellow marker in the top center. The code is plotting the differential evolution result as a cyan circle, but it is not visible on the plot. At a glance it's not clear which of these valleys is the true global minima."
    :include-source: 0
 
+Comparison of Global Optimizers
```
Comment (mdhaber) [inline doc/source/tutorial/optimize.rst:1009]:
```suggestion
Global optimization solver comparison
```
To match "Local minimization solver comparison"
But if that's the only thing, I won't change it here.

**Comment by mdhaber** [comment]
Looks accurate. I'll let @dschmitz89 close gh-17404 if this addresses all points. I think there is plenty more that can be done, but this is enough to merge. Thanks @koko1928 and reviewers!

**Comment by koko1928** [comment]
> Looks accurate. I'll let @dschmitz89 close [gh-17404](https://github.com/scipy/scipy/issues/17404) if this addresses all points. I think there was some information that [gh-17491](https://github.com/scipy/scipy/pull/17491) would have added that is not here, for instance. In any case, I think there is plenty more that can be done, but this is enough to merge. Thanks @koko1928 and reviewers!

Thank you!

**Pull request merged**